shared_model/builders/protobuf/builder_templates/proposal_template.hpp

```diff
@@ -0,0 +1,109 @@
+/**
+ * Copyright Soramitsu Co., Ltd. 2018 All Rights Reserved.
+ * http://soramitsu.co.jp
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef IROHA_PROTO_TEMPLATE_PROPOSAL_BUILDER_HPP
+#define IROHA_PROTO_TEMPLATE_PROPOSAL_BUILDER_HPP
+
+#include "backend/protobuf/proposal.hpp"
+#include "interfaces/common_objects/types.hpp"
+#include "validators/default_validator.hpp"
+
+#include "ordering.pb.h"
+
+namespace shared_model {
+  namespace proto {
+
+  /**
+     * Template proposal builder for creating new types of proposal builders by
+     * means of replacing template parameters
+     * @tparam S -- field counter for checking that all required fields are set
+     * @tparam SV -- stateless validator called when build method is invoked
+     */
+    template <int S = 0, typename SV = validation::DefaultProposalValidator>
+    class TemplateProposalBuilder {
+     private:
+      template <class T>
+      using w = detail::PolymorphicWrapper<T>;
+      using TransactionContainer = std::vector<w<Transaction>>;
+
+      template <int, typename>
+      friend class TemplateProposalBuilder;
+
+      enum RequiredFields { Transactions, Height, CreatedTime, TOTAL };
+
+      template <int s>
+      using NextBuilder = TemplateProposalBuilder<S | (1 << s), SV>;
+
+      iroha::ordering::proto::Proposal proposal_;
+      SV stateless_validator_;
+
+      template <int Sp, typename SVp>
+      TemplateProposalBuilder(const TemplateProposalBuilder<Sp, SVp> &o)
```

Comment:
Probably it should accept only `TemplateProposalBuilder<Sp, SV>`, so that there are no unexpected conversions with different validators.